### ISSUE TITLE
VALA

### ISSUE THREAD

**Issue by rainwoodman** [NumCosmo/NumCosmo#17]
There is this cute C# like programming called VALA that may simplify defining GObject classes. It compiles to C that calls the GLib Runtime. 

https://wiki.gnome.org/Projects/Vala

**Comment by rainwoodman** [comment]
I used it a few years ago on a medium sized project, (gnome-globalmenu) and had only good words for it. Now the compiler must have been more mature than then.

**Comment by vitenti** [comment]
I've been from time to time visiting Vala, I really liked it. I included the vapi as an output when the vala compiler is available. My problem is that I didn't find a way to mix C and Vala (at least in a way that I like), that way I couldn't define an object in Vala and extend in C *in the same project*. 

As I understand I would have to produce the vapi by hand or a much more complex procedure. Generate the bindings for the already written C+GObject, then compile the Vala written objects (which depend on those initial objects), if some of these Vala objects were to be extended in C, then produce another intermediary lib to build those and finally producing the library with all objects.

Of course, nothing stops anyone from using NumCosmo from Vala, but as part of the core development I can't see how to introduce it without introducing too much complexity.

**Comment by rainwoodman** [comment]
That's true. I was imagining a core written in C, with a vala binding, then the peripherals are written in Vala. But if most of the NumCosmo is a tightly coupled core, (it was mentioned it is hard to split the package to pieces) then there is no easy we to do this. I remember seeing from time to time people propose rewrite parts of GTK with VALA and then most of these efforts goes away. 

It would be nice to have some more elaborated examples and add a link to GNOME wiki. If I get around to play with this I'll make some. We can for example write a cosmology calculator app for gnome -- is NumCosmo packaged for major distributions?

**Comment by vitenti** [comment]
Yes, you can find some packages in https://build.opensuse.org/project/show/home:vitenti . We are preparing the next release 0.14.0, after it is done I'll update the packages. Some time ago I also made a macport for it but never published or updated it. 

In this new release I'm including much more detailed examples, including pure-python usage (models and data defined directly in python), Planck+CLASS (through NumCosmo CLASS frontend NcCBE), cross correlations, cluster cosmology, etc. 

The sad part of the python usage is that the current python implementations uses Gil (global interpreter lock) which turns it impractical for NumCosmo's multi-thread MCMC. For this reason I may include some examples in Perl or Vala.

A calculator app for GNOME would be great, I thought about doing it in Vala or Python.

**Comment by rainwoodman** [comment]
Thanks for the link. Looks not hard to add RPM package to Fedora/RHEL (I use Fedora). I can look into that at some point (e.g. when 0.14 is out?). My hands are tied currently into several other projects. 

I am not sure if I understand how GIL interferes with NumCosmo's multi-thread MCMC. 
Is it because NumCosmo has to call Python likelihood functions inside the MCMC chains?

**Comment by vitenti** [comment]
Up to now all likelihoods (NcmData) objects were written in C, but I've added the possibility of including additional likelihoods, implemented in any bound language, for a joint analysis. In the case that at least one of the NcmData objects were implemented in Python, the MCMC would call Python from different threads and would be locked by Gil. I have some ideas on how to solve this but no time right now to try them.
